File: androidApp/core/database/src/main/java/com/grateful/deadly/core/database/dao/FavoritesDao.kt
package com.grateful.deadly.core.database.dao

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import com.grateful.deadly.core.database.entities.FavoriteShowEntity
import kotlinx.coroutines.flow.Flow

/**
 * Favorites DAO - Database access for favorite show operations
 *
 * Provides reactive Flow-based queries and full CRUD operations
 * for user's favorites management with pin support and sorting.
 */
@Dao
interface FavoritesDao {

    // Core CRUD operations
    //
    // UI reads filter `deletedAt IS NULL`. Sync paths that need to see
    // tombstones use the *IncludingTombstones variants.

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun addToFavorites(favoriteShow: FavoriteShowEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun addMultipleToFavorites(favoriteShows: List<FavoriteShowEntity>)

    /** Soft-delete: row stays as a tombstone so sync can propagate the removal. */
    @Query("UPDATE favorite_shows SET deletedAt = :now, updatedAt = :now WHERE showId = :showId")
    suspend fun softDelete(showId: String, now: Long)

    @Update
    suspend fun updateFavoriteShow(favoriteShow: FavoriteShowEntity)

    // Reactive queries for UI
    @Query("SELECT * FROM favorite_shows WHERE deletedAt IS NULL ORDER BY isPinned DESC, addedToFavoritesAt DESC")
    fun getAllFavoriteShowsFlow(): Flow<List<FavoriteShowEntity>>

    @Query("SELECT * FROM favorite_shows WHERE isPinned = 1 AND deletedAt IS NULL ORDER BY addedToFavoritesAt DESC")
    fun getPinnedFavoriteShowsFlow(): Flow<List<FavoriteShowEntity>>

    @Query("SELECT * FROM favorite_shows WHERE deletedAt IS NULL ORDER BY isPinned DESC, addedToFavoritesAt DESC")
    suspend fun getAllFavoriteShows(): List<FavoriteShowEntity>

    // Individual show queries
    @Query("SELECT * FROM favorite_shows WHERE showId = :showId AND deletedAt IS NULL")
    suspend fun getFavoriteShowById(showId: String): FavoriteShowEntity?

    /** Sync helper: returns even tombstoned rows. */
    @Query("SELECT * FROM favorite_shows WHERE showId = :showId")
    suspend fun getFavoriteShowByIdIncludingTombstones(showId: String): FavoriteShowEntity?

    @Query("SELECT * FROM favorite_shows WHERE showId = :showId AND deletedAt IS NULL")
    fun getFavoriteShowByIdFlow(showId: String): Flow<FavoriteShowEntity?>

    // Status checks
    @Query("SELECT EXISTS(SELECT 1 FROM favorite_shows WHERE showId = :showId AND deletedAt IS NULL)")
    suspend fun isShowFavorite(showId: String): Boolean

    @Query("SELECT EXISTS(SELECT 1 FROM favorite_shows WHERE showId = :showId AND deletedAt IS NULL)")
    fun isShowFavoriteFlow(showId: String): Flow<Boolean>

    @Query("SELECT EXISTS(SELECT 1 FROM favorite_shows WHERE showId = :showId AND isPinned = 1 AND deletedAt IS NULL)")
    suspend fun isShowPinned(showId: String): Boolean

    @Query("SELECT EXISTS(SELECT 1 FROM favorite_shows WHERE showId = :showId AND isPinned = 1 AND deletedAt IS NULL)")
    fun isShowPinnedFlow(showId: String): Flow<Boolean>

    // Pin management
    @Query("UPDATE favorite_shows SET isPinned = :isPinned WHERE showId = :showId")
    suspend fun updatePinStatus(showId: String, isPinned: Boolean)

    @Query("UPDATE favorite_shows SET notes = :notes WHERE showId = :showId")
    suspend fun updateNotes(showId: String, notes: String?)

    @Query("UPDATE favorite_shows SET customRating = :rating WHERE showId = :showId")
    suspend fun updateCustomRating(showId: String, rating: Float?)

    @Query("UPDATE favorite_shows SET recordingQuality = :quality WHERE showId = :showId")
    suspend fun updateRecordingQuality(showId: String, quality: Int?)

    @Query("UPDATE favorite_shows SET playingQuality = :quality WHERE showId = :showId")
    suspend fun updatePlayingQuality(showId: String, quality: Int?)

    // Statistics
    @Query("SELECT COUNT(*) FROM favorite_shows WHERE deletedAt IS NULL")
    suspend fun getFavoriteShowCount(): Int

    @Query("SELECT COUNT(*) FROM favorite_shows WHERE isPinned = 1 AND deletedAt IS NULL")
    fun getPinnedShowCountFlow(): Flow<Int>

    @Query("SELECT COUNT(*) FROM favorite_shows WHERE deletedAt IS NULL")
    fun getFavoriteShowCountFlow(): Flow<Int>

    // Download tracking
    @Query("UPDATE favorite_shows SET downloadedRecordingId = :recordingId, downloadedFormat = :format WHERE showId = :showId")
    suspend fun updateDownloadedRecording(showId: String, recordingId: String?, format: String?)

    @Query("SELECT downloadedRecordingId FROM favorite_shows WHERE showId = :showId")
    suspend fun getDownloadedRecordingId(showId: String): String?

    // Bulk operations
    @Query("DELETE FROM favorite_shows")
    suspend fun clearFavorites()

    @Query("UPDATE favorite_shows SET isPinned = 0")
    suspend fun unpinAllShows()
}

File: iosApp/deadly/Core/Database/DAOs/FavoritesDAO.swift
import Foundation
import GRDB

struct FavoritesDAO: Sendable {
    let database: AppDatabase

    // MARK: - CRUD

    /// Upsert. Clears any tombstone (re-favoriting after delete is a resurrection).
    /// Caller is responsible for setting `updatedAt` on the incoming record.
    func add(_ favoriteShow: FavoriteShowRecord) throws {
        try database.write { db in
            var record = favoriteShow
            record.deletedAt = nil
            try record.insert(db, onConflict: .replace)
        }
    }

    /// Sync-apply upsert. Writes the record verbatim — does NOT clear the
    /// tombstone the way [add] does. Used by [UserSyncApplyService] so that
    /// a server tombstone propagates to the local DB.
    func applyFromSync(_ favoriteShow: FavoriteShowRecord) throws {
        try database.write { db in
            var record = favoriteShow
            try record.insert(db, onConflict: .replace)
        }
    }

    func addAll(_ favoriteShows: [FavoriteShowRecord]) throws {
        try database.write { db in
            for favoriteShow in favoriteShows {
                var record = favoriteShow
                record.deletedAt = nil
                try record.insert(db, onConflict: .replace)
            }
        }
    }

    /// Soft-delete: marks the row with deleted_at so sync can propagate the
    /// removal across devices. UI queries filter these out.
    func remove(_ showId: String) throws {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        try database.write { db in
            try FavoriteShowRecord
                .filter(Column("showId") == showId)
                .updateAll(db,
                    Column("deletedAt").set(to: now),
                    Column("updatedAt").set(to: now)
                )
        }
    }

    func update(_ favoriteShow: FavoriteShowRecord) throws {
        try database.write { db in
            var record = favoriteShow
            try record.update(db)
        }
    }

    // MARK: - Fetch
    //
    // All UI-facing reads filter out tombstones (deletedAt IS NULL).
    // Sync code paths that need to see tombstones should use
    // fetchAllIncludingTombstones / fetchByIdIncludingTombstones.

    private static let live = Column("deletedAt") == nil

    func fetchAll() throws -> [FavoriteShowRecord] {
        try database.read { db in
            try FavoriteShowRecord
                .filter(Self.live)
                .order(Column("isPinned").desc, Column("addedToFavoritesAt").desc)
                .fetchAll(db)
        }
    }

    func fetchAllIncludingTombstones() throws -> [FavoriteShowRecord] {
        try database.read { db in
            try FavoriteShowRecord.fetchAll(db)
        }
    }

    func fetchPinned() throws -> [FavoriteShowRecord] {
        try database.read { db in
            try FavoriteShowRecord
                .filter(Column("isPinned") == true && Self.live)
                .order(Column("addedToFavoritesAt").desc)
                .fetchAll(db)
        }
    }

    func fetchById(_ showId: String) throws -> FavoriteShowRecord? {
        try database.read { db in
            try FavoriteShowRecord
                .filter(Column("showId") == showId && Self.live)
                .fetchOne(db)
        }
    }

    /// Sync helper: returns the row even if tombstoned. Use when pushing
    /// pending changes so the outbox sees the deleted_at marker.
    func fetchByIdIncludingTombstones(_ showId: String) throws -> FavoriteShowRecord? {
        try database.read { db in
            try FavoriteShowRecord.fetchOne(db, key: showId)
        }
    }

    func isFavorite(_ showId: String) throws -> Bool {
        try database.read { db in
            try FavoriteShowRecord
                .filter(Column("showId") == showId && Self.live)
                .fetchCount(db) > 0
        }
    }

    func fetchCount() throws -> Int {
        try database.read { db in
            try FavoriteShowRecord.filter(Self.live).fetchCount(db)
        }
    }

    // MARK: - Pin management

    func updatePinStatus(_ showId: String, isPinned: Bool) throws {
        try database.write { db in
            try FavoriteShowRecord
                .filter(Column("showId") == showId)
                .updateAll(db, Column("isPinned").set(to: isPinned))
        }
    }

    // MARK: - Review fields

    func updateNotes(_ showId: String, notes: String?) throws {
        try database.write { db in
            try FavoriteShowRecord
                .filter(Column("showId") == showId)
                .updateAll(db, Column("notes").set(to: notes))
        }
    }

    func updateCustomRating(_ showId: String, rating: Double?) throws {
        try database.write { db in
            try FavoriteShowRecord
                .filter(Column("showId") == showId)
                .updateAll(db, Column("customRating").set(to: rating))
        }
    }

    func updateRecordingQuality(_ showId: String, quality: Int?) throws {
        try database.write { db in
            try FavoriteShowRecord
                .filter(Column("showId") == showId)
                .updateAll(db, Column("recordingQuality").set(to: quality))
        }
    }

    func updatePlayingQuality(_ showId: String, quality: Int?) throws {
        try database.write { db in
            try FavoriteShowRecord
                .filter(Column("showId") == showId)
                .updateAll(db, Column("playingQuality").set(to: quality))
        }
    }

    // MARK: - Management

    func clearAll() throws {
        try database.write { db in
            try FavoriteShowRecord.deleteAll(db)
        }
    }

    // MARK: - Observation

    func observeAll() -> ValueObservation<ValueReducers.Fetch<[FavoriteShowRecord]>> {
        ValueObservation.tracking { db in
            try FavoriteShowRecord
                .filter(Self.live)
                .order(Column("isPinned").desc, Column("addedToFavoritesAt").desc)
                .fetchAll(db)
        }
    }

    func observeCount() -> ValueObservation<ValueReducers.Fetch<Int>> {
        ValueObservation.tracking { db in
            try FavoriteShowRecord.filter(Self.live).fetchCount(db)
        }
    }

    // MARK: - Download tracking

    func updateDownloadedRecording(_ showId: String, recordingId: String?, format: String?) throws {
        try database.write { db in
            try FavoriteShowRecord
                .filter(Column("showId") == showId)
                .updateAll(db,
                    Column("downloadedRecordingId").set(to: recordingId),
                    Column("downloadedFormat").set(to: format)
                )
        }
    }

    func fetchDownloadedRecordingId(_ showId: String) throws -> String? {
        try database.read { db in
            try FavoriteShowRecord.fetchOne(db, key: showId)?.downloadedRecordingId
        }
    }

    func fetchShowsWithDownloads() throws -> [FavoriteShowRecord] {
        try database.read { db in
            try FavoriteShowRecord
                .filter(Column("downloadedRecordingId") != nil && Self.live)
                .fetchAll(db)
        }
    }

    func clearDownloadedRecording(_ showId: String) throws {
        try updateDownloadedRecording(showId, recordingId: nil, format: nil)
    }
}
